examples/rifle_sxiv.sh

```diff
@@ -11,34 +11,26 @@
 #   mime ^image, has sxiv, X, flag f = path/to/this/script -- "$@"
 #
 # Implementation notes: this script is quite slow because of POSIX limitations
-# and portability concerns. First calling the shell function 'abspath' is
+# and portability concerns. First, using case statement to get absolute path is
 # quicker than calling 'realpath' because it would fork a whole process, which
 # is slow. Second, we need to append a file list to sxiv, which can only be done
 # properly in two ways: arrays (which are not POSIX) or \0 sperated
 # strings. Unfortunately, assigning \0 to a variable is not POSIX either (will
 # not work in dash and others), so we cannot store the result of listfiles to a
 # variable.
 
-if [ $# -eq 0 ]; then
-    echo "Usage: ${0##*/} PICTURES"
-    exit
-fi
-
-[ "$1" = '--' ] && shift
-
-abspath () {
-    case "$1" in
-        /*) printf "%s\n" "$1";;
-        *)  printf "%s\n" "$PWD/$1";;
-    esac
-}
-
 listfiles () {
-    find -L "$(dirname "$target")" -maxdepth 1 -type f -iregex \
-      '.*\(jpe?g\|bmp\|png\|gif\)$' -print0 | sort -z
+    find -L "${target%/*}" -maxdepth 1 -type f -iregex \
```

Comment:
I think there's one edge-case here. If `target=/blah` then this'll expand to the empty string. Maybe `/${target%/*}`?

Reply:
Is having an extra leading `/` valid? If so then we can do what you suggested. Otherwise we can do this
```sh
dir="${target%/*}"
find -L "${dir:-/}"
```

Reply:
That was another thing I thought of. What if the user passes a directory *including* the trailing slash? `some/blah/`, for example, what's the desired behavior here?

I think a double slash is pretty universal, if triple is actually POSIX I guess it's better but I like the trailing slash idea fine too.

Oh, and another thing, doesn't the expansion result in the wrong result for `foo/bar/..`?

Reply:
>  What if the user passes a directory _including_ the trailing slash? `some/blah/`, for example, what's the desired behavior here?

On which case?
- If we use three leading slash then the trailing slash will be removed by the parameter expansion and everything works as expected. `////some/blah`
- If we use a single trailing slash then the parameter expansion will remove the trailing slash, but then the hard-coded one will be added in leading to everything working as expected as well.



> Oh, and another thing, doesn't the expansion result in the wrong result for `foo/bar/..`?

I don't understand this. Are you talking about `$target` not being an absolute path or about user inputting literal `..` ?

The first one is not exactly possible since the case statement will prevent that. And if it's a literal `..` then nothing changes from the previous implementation.

```sh
count="$(listfiles | grep -m 1 -ZznF "$target" | cut -d: -f1)"
```
Grep will try to find the `$target` and fail. `$count` will be empty leading to the fallback getting executed.
```sh
if [ -n "$count" ]; then
    listfiles | xargs -0 sxiv -n "$count" --
else
    sxiv -- "$@" # fallback
fi
```

Note: Sxiv can deal with trailing `..` so everything ends up working out as expected and no differently than the previous script.

Reply:
Not in any case with leading or trailing slashes. `foo/bar` results in `foo`, `foo/bar/` results in `foo/bar`, don't we want these almost identical paths to have the same result?

The relativeness of the path isn't the point. `foo/bar/..` is expanded to `foo/bar` but surely the user intends `foo/bar/..` to mean the same thing as `foo`? We don't really care about feature parity with the previous implementation if that was already wrong.

Reply:
> `foo/bar/..` is expanded to `foo/bar` but surely the user intends `foo/bar/..` to mean the same thing as `foo`?

That's exactly what happens though. The grep will fail, ``$count`` will be empty and the fallback will be executed, i.e ``sxiv -- foo/bar/..`` and sxiv will open the directory `foo` just as expected.

Reply:
https://github.com/ranger/ranger/pull/2411/commits/c47c0f3216fddedfb8099adbb98eece41f8ec040 should solve the edge case of file being in root. And I've tested out all the other possibilities you've listed and they all work as expected.

If user doesn't enter an ***image file*** (or more precisely a file with one of the extension ``find`` is listing out) as argument then grep would always fail and the fallback will be executed. If it's a directory (even with `..` in it, either trailing or in middle) then sxiv will open it as expected.

Reply:
Doesn't the find look *in* the stripped dir?

Reply:
> Doesn't the find look _in_ the stripped dir?

Yes, it lists out all the images in that dir. And then that list gets grepped for `"$target"`, if grep doesn't find anything then `$count` is not set and fallback is executed.

One edge case I can think of is if user enters ``foo/bar/../bar`` in this case the dir `bar` should be opened, but if the dir `foo` contains any image file starting with `bar`, grep will take that as a match and open up `foo/bar.png` instead of `foo/bar` the dir.

Reply:
Not sure if this is practical to fix without calling something like "realpath." Since handling of `..` in the presence of symlinks differs on Windows, for example. Which I *think* is also supposedly POSIX.

Reply:
The main problem here is that the grep command is not very strict.

This causes another interesting problem. If you enter `foo/b`, listfiles will list out all the images inside `foo`, and if there exists an image starting with `b` lets say `bar.png`, grep will take that as a match and open it up.

```diff
-grep -m 1 -ZznF "$target"
+grep -m 1 -Zzne "$target$"
```
This solves the issue, but I'm not sure what will happen if `$target` contains special chars.

Also speaking of symlinks and POSIX, I recently tried to find out a POSIX complaint way to resolve a symlink for a script of mine since `realpath` and `readlink` aren't POSIX. And at the end the most _""sane""_ way I found was parsing the output of ls.
```sh
resolved="$(ls -dl "${1}")"
echo "${resolved#*"$1" -> }"
```

Reply:
Just checking in to see whether you'd like to work on this further to deal with the edge cases? If not I'm fine with merging this as-is since the current behavior is retained.

Reply:
I don't mind, but I'm unable to think of a sane way to deal with this. Changing grep from fixed string to regex possibly opens up more bugs due to special chars in filename.